### PR TITLE
Implement virtual scroll for large dataframes

### PR DESCRIPTION
**Pull request by severo**
It replaces #347 (the remaining parts). The current commit will be recreated later, but I wanted to save my work.

remaining:
- [x] fix the update loop (lift some state up?)
- [x] when navigating down, the header row is sometimes focused instead of the next cell
- [x] slow: too many renders
- [x] check that everything was copied from #347 (CSS? mock row label location in the DOM?)
- [x] fix bottom border of the last row (see https://github.com/hyparam/hightable/pull/369#issuecomment-3718971157)
- [x] add tests (with playwright / vitest browser?)

Later?
- [ ] handle edge cases (eg. row / header height > viewport height)

https://github.com/user-attachments/assets/1a87d88d-c52c-4259-a08b-9b06eff9b85f

### CODE REVIEW

File: src/providers/ScrollModeVirtualProvider.tsx [hunk 1]
```diff
@@ -1,27 +1,219 @@
...
+    setVirtualScrollTop(toVirtualScrollTop(scrollTop))
+  } else if (Math.abs(difference) > finePrecision) {
+    // scrollTop changed slightly, virtualScrollTop controls the position (fine scroll)
+    // don't change virtualScrollTop, since it is more precise
+    // and set scrollTop back to the precise value
+    scrollToTop?.(expectedScrollTop, { behavior: 'instant' })
+    // TODO(SL): one issue is that when scrolling with a key arrow and the whole table is focused
+    // (not a cell), the scroll behavior is weird, it sometimes catches, and sometimes it is reverted.
+    // Maybe adding some throttle on scroll events would help
+  }
```
[the remark this answers is not in the document]
Comment (severo) [reply src/providers/ScrollModeVirtualProvider.tsx]: no

File: src/providers/ScrollModeVirtualProvider.tsx [hunk 2]
```diff
@@ -1,27 +1,219 @@
...
+  [firstVisibleRow, numRows, virtualScrollTop, clientHeight, headerHeight])
+
+  // const hiddenPixelsAfter = headerHeight + (lastVisibleRow + 1) * rowHeight - (virtualScrollTop + clientHeight)
+
+  if (isNaN(firstVisibleRow)) throw new Error(`invalid start row ${firstVisibleRow}`)
+  if (isNaN(lastVisibleRow)) throw new Error(`invalid end row ${lastVisibleRow + 1}`)
+  if (lastVisibleRow + 1 - firstVisibleRow > 1000) throw new Error(`attempted to render too many rows ${lastVisibleRow + 1 - firstVisibleRow}`)
+  if (visibleRowsRange?.start !== firstVisibleRow || visibleRowsRange.end !== lastVisibleRow + 1) {
+    setVisibleRowsRange?.({ start: firstVisibleRow, end: lastVisibleRow + 1 })
+  }
```
[the remark this answers is not in the document]
Comment (severo) [reply src/providers/ScrollModeVirtualProvider.tsx]: no

File: src/providers/ScrollModeVirtualProvider.tsx [hunk 3]
```diff
@@ -1,27 +1,219 @@
...
+
+  // scrollTop has a limited precision (1px, or subpixel on some browsers) and is not predictable exactly, in particular when used with zooming.
+  // The browser might also scroll slightly when focusing an element.
+  // preciseScrollTop is decimal, computed for the virtual canvas, and updated by user action only when scrollTop changes significantly.
+  const coarsePrecision = 40 // px. TODO(SL): how to choose the value? in percentage of clientHeight/canvasHeight? Generally, it's 33px or -33px (the height of the focused cell?)
+  const finePrecision = 1
+  const expectedScrollTop = useMemo(() => toScrollTop(virtualScrollTop), [toScrollTop, virtualScrollTop])
+  const difference = useMemo(() => scrollTop - expectedScrollTop, [scrollTop, expectedScrollTop])
+
+  if (Math.abs(difference) > coarsePrecision) {
```
[the remark this answers is not in the document]
Comment (severo) [reply src/providers/ScrollModeVirtualProvider.tsx]: ok

File: src/providers/ScrollModeVirtualProvider.tsx [hunk 4]
```diff
@@ -1,27 +1,219 @@
...
+    // scrollTop changed significantly, it controls the position (coarse scroll)
+    setVirtualScrollTop(toVirtualScrollTop(scrollTop))
+  } else if (Math.abs(difference) > finePrecision) {
```
[the remark this answers is not in the document]
Comment (severo) [reply src/providers/ScrollModeVirtualProvider.tsx]: ok

File: src/hooks/useCellFocus.ts [hunk 5]
```diff
@@ -16,35 +16,63 @@ interface CellFocus {
...
+          setShouldScrollHorizontally?.(false)
         }
-        element.focus()
-        setShouldFocus(false)
+      }, options)
+      observer.observe(element)
+
+      return () => {
+        observer.unobserve(element)
+        observer.disconnect()
```
[the remark this answers is not in the document]
Comment (severo) [reply src/hooks/useCellFocus.ts]: indeed, we return a cleanup function and use it in some components, but not in others. Fixed.

File: src/providers/ScrollModeVirtualProvider.tsx [hunk 6]
```diff
@@ -1,27 +1,219 @@
...
+    if (canvasHeight > clientHeight && scrollTop > canvasHeight - clientHeight) {
+      throw new Error(`Invalid scrollTop: ${scrollTop} when canvasHeight > clientHeight. It should be less than or equal to canvasHeight - clientHeight (${canvasHeight - clientHeight}, canvasHeight: ${canvasHeight}, clientHeight: ${clientHeight}).`)
+    }
+
+    // TODO(SL): remove this fallback (100px)? It's only for the tests in Node.js, where the elements have zero height
+    // instead, it should return without updating the visible rows range, or set it to undefined.
+    // TODO(SL): test in the browser (playwright)
+    setClientHeight(clientHeight === 0 ? 100 : clientHeight)
+    setScrollTop(scrollTop)
+  }, [canvasHeight])
```
[the remark this answers is not in the document]
Comment (severo) [reply src/providers/ScrollModeVirtualProvider.tsx]: indeed. Fixed

File: src/providers/ScrollModeVirtualProvider.tsx [hunk 7]
```diff
@@ -1,27 +1,219 @@
...
+   * Programmatically scroll to a specific row if needed.
+   * Beware:
+   * - row 1: header
+   * - row 2: first data row
+   * - row numRows + 1: last data row
+   * @param row The row to scroll to (same semantic as aria-rowindex: 1-based, includes header)
+   */
+  const scrollRowIntoView = useCallback(({ rowIndex }: { rowIndex: number }) => {
+    if (rowIndex < 1 || rowIndex > numRows + 2 || !Number.isInteger(rowIndex)) {
+      throw new Error(`Invalid first visible row index: ${rowIndex}. It should be an integer between 1 and ${numRows + 2}.`)
```
[the remark this answers is not in the document]
Comment (severo) [reply src/providers/ScrollModeVirtualProvider.tsx]: indeed

File: src/providers/ScrollModeVirtualProvider.tsx [hunk 8]
```diff
@@ -1,27 +1,219 @@
...
+
+  // convert scrollTop (in canvas coordinates, between 0px and canvasHeight - clientHeight)
+  // to virtualScrollTop (in virtual canvas coordinates, between 0px and headerHeight + numRows * rowHeight - clientHeight)
+  const toVirtualScrollTop = useCallback((scrollTop: number) => {
+    return scrollTop * (virtualCanvasHeight - clientHeight) / (canvasHeight - clientHeight)
+  }, [virtualCanvasHeight, clientHeight, canvasHeight])
+
+  // convert virtualScrollTop (in virtual canvas coordinates, between 0px and headerHeight + numRows * rowHeight - clientHeight)
+  // to scrollTop (in canvas coordinates, between 0px and canvasHeight - clientHeight)
+  const toScrollTop = useCallback((virtualScrollTop: number) => {
```
[the remark this answers is not in the document]
Comment (severo) [reply src/providers/ScrollModeVirtualProvider.tsx]: done

[the remark this answers is not in the document]
Comment (severo) [reply src/providers/ScrollModeVirtualProvider.tsx]: done

File: src/providers/ScrollModeVirtualProvider.tsx [hunk 9]
```diff
@@ -1,27 +1,219 @@
...
+    return virtualScrollTop * (canvasHeight - clientHeight) / (virtualCanvasHeight - clientHeight)
+  }, [canvasHeight, clientHeight, virtualCanvasHeight])
+
+  // scrollTop has a limited precision (1px, or subpixel on some browsers) and is not predictable exactly, in particular when used with zooming.
+  // The browser might also scroll slightly when focusing an element.
+  // preciseScrollTop is decimal, computed for the virtual canvas, and updated by user action only when scrollTop changes significantly.
+  const coarsePrecision = 40 // px. TODO(SL): how to choose the value? in percentage of clientHeight/canvasHeight? Generally, it's 33px or -33px (the height of the focused cell?)
```
[the remark this answers is not in the document]
Comment (severo) [reply src/providers/ScrollModeVirtualProvider.tsx]: there's a TODO

File: src/hooks/useCellFocus.ts [hunk 10]
```diff
@@ -16,35 +16,63 @@ interface CellFocus {
...
-          // scroll-padding-inline-start and scroll-padding-block-start are set in the CSS
-          // to avoid the cell being hidden by the row and column headers
-          //
-          // not applied for header cells, as they are always visible, and it was causing jumps when resizing a column
+      // from https://github.com/hyparam/hightable/pull/347
+      // TODO(SL): revisit this logic later
+      const options = {
+        rootMargin: '0px',
+        scrollMargin: '0px',
+        threshold: 1.0,
```
[the remark this answers is not in the document]
Comment (severo) [reply src/hooks/useCellFocus.ts]: there's a TODO

File: src/hooks/useCellFocus.ts [hunk 11]
```diff
@@ -10,41 +10,67 @@ interface CellData {
...
-          //
-          // scroll-padding-inline-start and scroll-padding-block-start are set in the CSS
-          // to avoid the cell being hidden by the row and column headers
-          //
-          // not applied for header cells, as they are always visible, and it was causing jumps when resizing a column
+      // from https://github.com/hyparam/hightable/pull/347
+      // TODO(SL): revisit this logic later
+      const options = {
+        rootMargin: '0px',
+        scrollMargin: '0px',
```
[the remark this answers is not in the document]
Comment (severo) [reply src/hooks/useCellFocus.ts]: false

File: src/providers/ScrollModeVirtualProvider.tsx [hunk 12]
```diff
@@ -1,27 +1,232 @@
...
+
+    // partly or totally hidden: update the scroll position
+    const delta = hiddenPixelsBefore > 0 ? -hiddenPixelsBefore : hiddenPixelsAfter
+    const newVirtualScrollTop = virtualScrollTop + delta
+
+    // Update the virtual scroll top
+    setVirtualScrollTop(newVirtualScrollTop)
+
+    const tolerancePixels = 1
+    if (Math.abs(toScrollTop(delta)) > tolerancePixels) {
```
[the remark this answers is not in the document]
Comment (severo) [reply src/providers/ScrollModeVirtualProvider.tsx]: right, it's better

File: src/hooks/useCellFocus.ts [hunk 13]
```diff
@@ -10,41 +10,67 @@ interface CellData {
...
+          // scroll to the cell only if it's not a header cell
+          // note: we want it to scroll only horizontally. Vertically, it should already be in view.
+          // TODO(SL): container: nearest could be useful, if we decide to have two scrollable containers (one for vertical, one for horizontal)
+          // but unfortunately it's not widely supported yet (no Firefox support as of January 2026)
           element.scrollIntoView({ behavior: 'auto', block: 'nearest', inline: 'nearest' })
+          setShouldScrollHorizontally?.(false)
         }
-        element.focus()
-        setShouldFocus(false)
+      }, options)
```
[the remark this answers is not in the document]
Comment (severo) [reply src/hooks/useCellFocus.ts]: ahah ok, I didn't know

[the remark this answers is not in the document]
Comment (severo) [reply src/hooks/useCellFocus.ts]: hmmm, the horizontal scrolling on large data is broken again...

[the remark this answers is not in the document]
Comment (severo) [reply src/hooks/useCellFocus.ts]: fixed

[the remark this answers is not in the document]
Comment (severo) [reply src/hooks/useCellFocus.ts]: it works without the IntersectionObserver, much better!

File: src/components/HighTable/Scroller.tsx [hunk 14]
```diff
@@ -42,7 +42,11 @@ export default function Scroller({
...
    */
-  const viewportRef = useCallback((viewport: HTMLDivElement) => {
+  const viewportRef = useCallback((viewport: HTMLDivElement | null) => {
+    if (!viewport) {
+      return
+    }
+
     // Use arrow functions to get correct viewport type (not null)
     // eslint-disable-next-line func-style
     const updateViewportSize = () => {
```
[the remark this answers is not in the document]
Comment (severo) [reply src/components/HighTable/Scroller.tsx:52]: ok

File: src/components/HighTable/Slice.tsx [hunk 15]
```diff
@@ -108,8 +108,8 @@ export default function Slice({
     event.preventDefault()
     if (newRowIndex !== undefined) {
       setRowIndex(newRowIndex)
-      scrollRowIntoView?.({ rowIndex: newRowIndex }) // ensure the cell is visible
     }
+    scrollRowIntoView?.({ rowIndex: newRowIndex ?? rowIndex }) // ensure the cell is visible
```
[the remark this answers is not in the document]
Comment (severo) [reply src/components/HighTable/Slice.tsx]: added a TODO

File: src/providers/ScrollModeVirtualProvider.tsx [hunk 16]
```diff
@@ -1,27 +1,227 @@
...
+  // convert scrollTop (in canvas coordinates, between 0px and canvasHeight - clientHeight)
+  // to virtualScrollTop (in virtual canvas coordinates, between 0px and headerHeight + numRows * rowHeight - clientHeight)
+  const toVirtualScrollTop = useCallback((scrollTop: number) => {
+    return scrollTop * (virtualCanvasHeight - clientHeight) / (canvasHeight - clientHeight)
+  }, [virtualCanvasHeight, clientHeight, canvasHeight])
+
+  // convert virtualScrollTop (in virtual canvas coordinates, between 0px and headerHeight + numRows * rowHeight - clientHeight)
+  // to scrollTop (in canvas coordinates, between 0px and canvasHeight - clientHeight)
+  const toScrollTop = useCallback((virtualScrollTop: number) => {
+    return virtualScrollTop * (canvasHeight - clientHeight) / (virtualCanvasHeight - clientHeight)
```
[the remark this answers is not in the document]
Comment (severo) [reply src/providers/ScrollModeVirtualProvider.tsx]: added  a check (should not occur)

**Comment by severo** [comment]
Finally, I have a valid PR for virtual scrolling of huge dataframes. Would you like to have a look @platypii?

**Comment by severo** [comment]
The bottom border is not visible when scrolling past the end of the table (it only happens in "virtual scroll mode")


https://github.com/user-attachments/assets/21ab8721-7962-4f21-8085-b6daacc7739d

**Comment by platypii** [comment]
Still getting weird behavior: if I scroll down even one "notch" on my scroll wheel, I jump to row 4000+. 

Worse: when I try to scroll back up, there's no way to get back to row 1

https://github.com/user-attachments/assets/0186bdcf-cd38-4eb1-86d5-b72fd543da57

File: src/providers/ScrollModeVirtualProvider.tsx [hunk 17]
```diff
@@ -1,27 +1,375 @@
...
+      virtualCanvasHeight,
+      clientHeight,
+    }
+  }, [clientHeight, canvasHeight, headerHeight, numRows])
+
+  // ideally: call SET_SCALE from an event listener (if num_rows changes, or on resize if clientHeight or headerHeight change)
+  // not during rendering
+  if (currentScale && currentScale !== scale) {
+    dispatch({ type: 'SET_SCALE', scale: currentScale })
+  }
```
[the remark this answers is not in the document]
Comment (severo) [reply src/providers/ScrollModeVirtualProvider.tsx:40]: it's better than useEffect

File: src/providers/ScrollModeVirtualProvider.tsx [hunk 18]
```diff
@@ -1,27 +1,375 @@
...
+    }
+
+    const row = rowIndex - ariaOffset // convert to 0-based data row index
+
+    // Three cases:
+    // - the row is fully visible: do nothing
+    // - the row start is before virtualScrollTop + headerHeight: scroll to snap its start with that value
+    // - the row end is after virtualScrollTop + viewportHeight: scroll to snap its end with that value
+    const virtualScrollTop = virtualScrollBase + virtualScrollDelta
+    const hiddenPixelsBefore = virtualScrollTop + headerHeight - (headerHeight + row * rowHeight)
```
[the remark this answers is not in the document]
Comment (severo) [reply src/providers/ScrollModeVirtualProvider.tsx]: no

**Review by platypii — COMMENTED** [review]
NICE!! This is feeling much better! 👏 

I noticed one weird behavior: if you scroll all the way to the bottom of the "Large Data" storybook you get "stuck" and need to scroll up a bunch to make it "unstick" from the bottom.

**Comment by severo** [comment]
Yes, I would have changed the second number if the first one was not zero. Until 1.0.0, I was thinking that changes that don't break anything must go to the third number. Anyway, if you prefer the second number for communication matters, I can release 0.26.0 too.

**Comment by severo** [comment]
I updated the demo: try it at https://hyparam.github.io/demos/hightable/#/large
